Homework 01/task 03

jmilic1 opened the pull request.

Commit:
Added task specification

Commit:
Added pokemon encounter fetching

Commit:
Added mod and sum files

Commit:
Fixed bug when fetching by id

Commit:
Added documentation

mbekavac-happening commented on homework_1/Task-03/main.go:
```diff
@@ -0,0 +1,119 @@
+)
+
+// url contains base path to pokeAPI
+const url = "https://pokeapi.co/api/v2/pokemon/"
```
API parsing specific things should not be within the main package - please extract them to a separate package

mbekavac-happening commented on homework_1/Task-03/main.go:
```diff
@@ -0,0 +1,119 @@
+// it is defined by json response for "https://pokeapi.co/api/v2/pokemon/1"
+type pokemon struct {
+	Name                     string
+	Location_area_encounters string
```
Please use [struct tags](https://www.notion.so/superbet/1st-homework-common-mistakes-comments-1b21acc934bb42f8b9676d7da7fd128c#6e3df6b3560740a8b9e6ded06c1c2993) instead of names and  keep the names consistent with the convention.

jmilic1 replied:
Can I somehow request access to this link? It says I need to contact an admin

mbekavac-happening replied:
Sorry, I thought the page was shared with you guys - the contents are copy pasted here - https://axilis.slack.com/archives/C020JHMJ274/p1620815275002100

jmilic1 replied:
Thanks!

mbekavac-happening commented on homework_1/Task-03/main.go:
```diff
@@ -0,0 +1,119 @@
+const url = "https://pokeapi.co/api/v2/pokemon/"
+
+// pokemonDto contains fields which need to be in output
+type pokemonDto struct {
```
If you didn't have access to the API or its docs, what do you say would be the difference between what you use the `pokemonDto` for and what you use `pokemon` for?

I suggest finding better names. Something like "pokemonData" and "encounterData" comes to mind for the API results + "PokemonEncounters" for the final struct.

mbekavac-happening commented on homework_1/Task-03/main.go:
```diff
@@ -0,0 +1,119 @@
+	var err error
+
+	if id == 0 {
+		bodyContent, err = fetchFromUrl(url + name)
```
please use proper and safe joining function https://stackoverflow.com/questions/34668012/combine-url-paths-with-path-join/34668130

mbekavac-happening commented on homework_1/Task-03/main.go:
```diff
@@ -0,0 +1,119 @@
+	var id int
+	var name string
+
+	flag.IntVar(&id, "id", 0, "Id of Pokemon to be retrieved")
```
different argument parsing than in the first task, please be consistent

mbekavac-happening commented on homework_1/Task-03/main.go:
```diff
@@ -0,0 +1,119 @@
+	var bodyContent []byte
+	var err error
+
+	if id == 0 {
```
since you're just doing url + passed parameter, regardless of whether an ID or a name is passed in, I think you don't have to make a distinction here and just let the pokeapi validate that for you

mbekavac-happening commented on homework_1/Task-03/main.go:
```diff
@@ -0,0 +1,119 @@
+
+// fetchFromUrl fetches data from given url.
+// returns data as an array of bytes and an error if it occurred
+func fetchFromUrl(url string) ([]byte, error) {
```
nice

mbekavac-happening commented on homework_1/Task-03/main.go:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	var decodedPokemon pokemon
+	err = json.Unmarshal(bodyContent, &decodedPokemon)
```
You're probably wondering if this turned out to be too verbose - it did but there's little you can do to help it. Go has a pretty verbose way of dealing with the errors. There are some tricks (probably covered on the next Saturday) to make it a bit shorter, but nothing really helps.

However, you should encapsulate all this and move it outside of the main function (and the main package) - treat main as a client - it shouldn't know of the details of how the API is designed and what needs to be fetched and how to merge everything etc.

Commit:
Removed task specification

Commit:
Moved logic from main program to separate pokemon-encounter-api component.

Merged "id" and "name" flags into a single "pokemon" flag of type string.

Renamed data models and their fields. Fields now use struct tags for json specification.

Refactored url concatenation to use path.Join method.

Removed unnecessary documentation in fetchFromUrl method.

Commit:
Added check to see if flag set the input variable to a nil pointer.

Commit:
Simplified name of pokemon-encounters package to.

Commit:
Removed gitkeep

mbekavac-happening commented on homework_1/Task-03/encounters/encounters-api.go:
```diff
@@ -0,0 +1,112 @@
+package encounters
```
camelCase is preferred over kebab-case for file names

mbekavac-happening replied:
also, this package doesn't only contain logic around the encounters, but encapsulates all the (necessary) pokeapi logic

mbekavac-happening replied:
One meta-comment - please don't resolve reviewer comments yourself - it makes it a bit more difficult to re-review (have to open the comments again, they aren't in the "Files changed" tab anymore etc.). It also means you're 100% certain you completely covered whatever was commented, which is not always the case. Don't worry about it, it's a piece of PR etiquette that isn't obvious and different teams have different conventions.

As for this particular comment (name of the package), I think you missed my point. Feel free to ask for clarification when in doubt. I was hinting at the fact that a more appropriate name would be `pokeapi` or `pokeapiclient` or something like that. You're not only writing logic for fetching encounters, you're writing logic for some other endpoints as well + merging all the fetched data. In a setting where this is not just a homework assignment, you could also extend this package to contain functions for other endpoints, transforming it into a proper PokeAPI client/wrapper.

Anyway, this is nitpicking so no changes necessary from my side.

mbekavac-happening reviewed (commented):
nice, same comment as on the task number 2

Commit:
Renamed 'encounters' to 'encountersApi'
